File: builtins/Math.js
// Self-hosted Math builtins on top of __ops__
var ops = __ops__;
var ThrowTypeError = ops.ThrowTypeError;

// Save native implementations before overriding
var _nativeFloor = Math.__floor;
var _nativeCeil = Math.__ceil;
var _nativeRound = Math.__round;
var _nativeTrunc = Math.__trunc;
var _nativeSign = Math.__sign;
var _nativePow = Math.__pow;
var _nativeSqrt = Math.__sqrt;

// Math.max (ES2025 §21.3.2.24)
Math.max = function MathMax() {
  var max = -Infinity;
  for (var i = 0; i < arguments.length; i++) {
    var n = Number(arguments[i]);
    if (n !== n) return n;
    if (n > max) max = n;
  }
  return max;
};

// Math.min (ES2025 §21.3.2.27)
Math.min = function MathMin() {
  var min = Infinity;
  for (var i = 0; i < arguments.length; i++) {
    var n = Number(arguments[i]);
    if (n !== n) return n;
    if (n < min) min = n;
  }
  return min;
};

// Math.abs (ES2025 §21.3.2.1)
Math.abs = function MathAbs(x) {
  var n = Number(x);
  return n < 0 ? -n : (n === 0 ? 0 : n);
};

// Math.floor (ES2025 §21.3.2.16)
Math.floor = function MathFloor(x) {
  return _nativeFloor(Number(x));
};

// Math.ceil (ES2025 §21.3.2.9)
Math.ceil = function MathCeil(x) {
  return _nativeCeil(Number(x));
};

// Math.round (ES2025 §21.3.2.30)
Math.round = function MathRound(x) {
  return _nativeRound(Number(x));
};

// Math.trunc (ES2025 §21.3.2.37)
Math.trunc = function MathTrunc(x) {
  return _nativeTrunc(Number(x));
};

// Math.sign (ES2025 §21.3.2.33)
Math.sign = function MathSign(x) {
  return _nativeSign(Number(x));
};

// Math.pow (ES2025 §21.3.2.29)
Math.pow = function MathPow(x, y) {
  return _nativePow(Number(x), Number(y));
};

// Math.sqrt (ES2025 §21.3.2.34)
Math.sqrt = function MathSqrt(x) {
  return _nativeSqrt(Number(x));
};

Math.sin = function(x) { return Math.__sin(Number(x)); };
Math.cos = function(x) { return Math.__cos(Number(x)); };
Math.tan = function(x) { return Math.__tan(Number(x)); };
Math.asin = function(x) { return Math.__asin(Number(x)); };
Math.acos = function(x) { return Math.__acos(Number(x)); };
Math.atan = function(x) { return Math.__atan(Number(x)); };
Math.log = function(x) { return Math.__log(Number(x)); };
Math.log10 = function(x) { return Math.__log10(Number(x)); };
Math.log2 = function(x) { return Math.__log2(Number(x)); };
Math.exp = function(x) { return Math.__exp(Number(x)); };
Math.log1p = function(x) { return Math.__log1p(Number(x)); };
Math.cbrt = function(x) { return Math.__cbrt(Number(x)); };
Math.expm1 = function(x) { return Math.__expm1(Number(x)); };
Math.cosh = function(x) { return Math.__cosh(Number(x)); };
Math.sinh = function(x) { return Math.__sinh(Number(x)); };
Math.tanh = function(x) { return Math.__tanh(Number(x)); };
Math.acosh = function(x) { return Math.__acosh(Number(x)); };
Math.asinh = function(x) { return Math.__asinh(Number(x)); };
Math.atanh = function(x) { return Math.__atanh(Number(x)); };
Math.atan2 = function(y, x) { return Math.__atan2(Number(y), Number(x)); };
Math.imul = function(x, y) { return Math.__imul(Number(x), Number(y)); };
Math.fround = function(x) { return Math.__fround(Number(x)); };
Math.clz32 = function(x) { return Math.__clz32(Number(x)); };
Math.hypot = function() { return Math.__hypot.apply(Math, arguments); };
Math.random = function() { return Math.__random(); };

File: crates/quench-runtime/src/builtin_meta/math.rs
//! Math builtin metadata.

use crate::ops::Builtin;

pub const fn fn_name(b: Builtin) -> Option<&'static str> {
    match b {
        Builtin::MathAbs => Some("Math.abs"),
        Builtin::MathFloor => Some("Math.floor"),
        Builtin::MathCeil => Some("Math.ceil"),
        Builtin::MathRound => Some("Math.round"),
        Builtin::MathTrunc => Some("Math.trunc"),
        Builtin::MathMax => Some("Math.max"),
        Builtin::MathMin => Some("Math.min"),
        Builtin::MathSign => Some("Math.sign"),
        Builtin::MathSqrt => Some("Math.sqrt"),
        Builtin::MathCbrt => Some("Math.cbrt"),
        Builtin::MathHypot => Some("Math.hypot"),
        Builtin::MathImul => Some("Math.imul"),
        Builtin::MathLog => Some("Math.log"),
        Builtin::MathLog10 => Some("Math.log10"),
        Builtin::MathLog2 => Some("Math.log2"),
        Builtin::MathExp => Some("Math.exp"),
        Builtin::MathSin => Some("Math.sin"),
        Builtin::MathCos => Some("Math.cos"),
        Builtin::MathTan => Some("Math.tan"),
        Builtin::MathAsin => Some("Math.asin"),
        Builtin::MathAcos => Some("Math.acos"),
        Builtin::MathAtan => Some("Math.atan"),
        _ => fn_name_tail(builtin),
    }
}

const fn fn_name_tail(builtin: Builtin) -> Option<&'static str> {
    match builtin {
        Builtin::MathAtan2 => Some("Math.atan2"),
        Builtin::MathAcosh => Some("Math.acosh"),
        Builtin::MathAsinh => Some("Math.asinh"),
        Builtin::MathAtanh => Some("Math.atanh"),
        Builtin::MathClz32 => Some("Math.clz32"),
        Builtin::MathCosh => Some("Math.cosh"),
        Builtin::MathExpm1 => Some("Math.expm1"),
        Builtin::MathFround => Some("Math.fround"),
        Builtin::MathLog1p => Some("Math.log1p"),
        Builtin::MathSinh => Some("Math.sinh"),
        Builtin::MathTanh => Some("Math.tanh"),
        Builtin::MathF16Round => Some("Math.f16round"),
        Builtin::MathRandom => Some("Math.random"),
        Builtin::MathSumPrecise => Some("Math.sumPrecise"),
        Builtin::MathPow => Some("Math.pow"),
        _ => None,
    }
}

pub const fn fn_len(b: Builtin) -> Option<f64> {
    match b {
        Builtin::MathAtan2
        | Builtin::MathHypot
        | Builtin::MathImul
        | Builtin::MathMax
        | Builtin::MathMin
        | Builtin::MathPow => Some(2.0),
        Builtin::MathAbs
        | Builtin::MathFloor
        | Builtin::MathCeil
        | Builtin::MathRound
        | Builtin::MathTrunc
        | Builtin::MathSign
        | Builtin::MathSqrt
        | Builtin::MathCbrt
        | Builtin::MathLog
        | Builtin::MathLog10
        | Builtin::MathLog2
        | Builtin::MathExp
        | Builtin::MathSin
        | Builtin::MathCos
        | Builtin::MathTan
        | Builtin::MathAsin
        | Builtin::MathAcos
        | Builtin::MathAtan
        | Builtin::MathAcosh
        | Builtin::MathAsinh
        | Builtin::MathAtanh
        | Builtin::MathClz32
        | Builtin::MathCosh
        | Builtin::MathExpm1
        | Builtin::MathFround
        | Builtin::MathLog1p
        | Builtin::MathSinh
        | Builtin::MathTanh => Some(1.0),
        _ => fn_len_tail(b),
    }
}

const fn fn_len_tail(b: Builtin) -> Option<f64> {
    match b {
        Builtin::MathF16Round | Builtin::MathSumPrecise => Some(1.0),
        Builtin::MathRandom => Some(0.0),
        _ => None,
    }
}

pub const fn short_name(b: Builtin) -> Option<&'static str> {
    match b {
        Builtin::MathAbs => Some("abs"),
        Builtin::MathFloor => Some("floor"),
        Builtin::MathCeil => Some("ceil"),
        Builtin::MathRound => Some("round"),
        Builtin::MathTrunc => Some("trunc"),
        Builtin::MathMax => Some("max"),
        Builtin::MathMin => Some("min"),
        Builtin::MathSign => Some("sign"),
        Builtin::MathSqrt => Some("sqrt"),
        Builtin::MathCbrt => Some("cbrt"),
        Builtin::MathHypot => Some("hypot"),
        Builtin::MathImul => Some("imul"),
        Builtin::MathLog => Some("log"),
        Builtin::MathLog10 => Some("log10"),
        Builtin::MathLog2 => Some("log2"),
        Builtin::MathExp => Some("exp"),
        Builtin::MathSin => Some("sin"),
        Builtin::MathCos => Some("cos"),
        Builtin::MathTan => Some("tan"),
        Builtin::MathAsin => Some("asin"),
        Builtin::MathAcos => Some("acos"),
        Builtin::MathAtan => Some("atan"),
        Builtin::MathAtan2 => Some("atan2"),
        Builtin::MathAcosh => Some("acosh"),
        Builtin::MathAsinh => Some("asinh"),
        Builtin::MathAtanh => Some("atanh"),
        Builtin::MathClz32 => Some("clz32"),
        Builtin::MathCosh => Some("cosh"),
        Builtin::MathExpm1 => Some("expm1"),
        Builtin::MathFround => Some("fround"),
        Builtin::MathLog1p => Some("log1p"),
        Builtin::MathSinh => Some("sinh"),
        Builtin::MathTanh => Some("tanh"),
        Builtin::MathF16Round => Some("f16round"),
        Builtin::MathRandom => Some("random"),
        Builtin::MathSumPrecise => Some("sumPrecise"),
        Builtin::MathPow => Some("pow"),
        _ => None,
    }
}
